lib/kube/proxy/auth.go

```diff
@@ -121,57 +125,68 @@ func getKubeCreds(ctx context.Context, log logrus.FieldLogger, tpClusterName, ku
 	res := make(map[string]*kubeCreds, len(cfg.Contexts))
 	// Convert kubeconfig contexts into kubeCreds.
 	for cluster, clientCfg := range cfg.Contexts {
-		log := log.WithField("cluster", cluster)
-		log.Debug("Checking Kubernetes impersonation permissions.")
-		client, err := kubernetes.NewForConfig(clientCfg)
+		clusterCreds, err := extractKubeCreds(ctx, cluster, clientCfg, serviceType, kubeconfigPath, log)
 		if err != nil {
-			return nil, trace.Wrap(err, "failed to generate Kubernetes client for cluster %q", cluster)
-		}
-		// For each loaded cluster, check impersonation permissions. This
-		// failure is only critical for newKubeService.
-		if err := checkImpersonationPermissions(ctx, client.AuthorizationV1().SelfSubjectAccessReviews()); err != nil {
-			// kubernetes_service must have valid RBAC permissions, otherwise
-			// it's pointless.
-			// proxy_service can run without them (e.g. a root proxy).
-			if serviceType == KubeService {
-				return nil, trace.Wrap(err)
-			}
-			log.WithError(err).Warning("Failed to test the necessary Kubernetes permissions. This teleport instance will still handle Kubernetes requests towards other Kubernetes clusters")
-			// We used to recommend users to set a dummy kubeconfig on root
-			// proxies to get kubernetes support working for leaf clusters:
-			// https://community.goteleport.com/t/enabling-teleport-to-act-as-a-kubernetes-proxy-for-trusted-leaf-clusters/418
-			//
-			// Since this is no longer necessary, recommend them to clean up
-			// via logs.
-			if kubeconfigPath != "" {
-				log.Info("If this is a proxy and you provided a dummy kubeconfig_path, you can remove it from teleport.yaml to get rid of this warning")
-			}
-		} else {
-			log.Debug("Have all necessary Kubernetes impersonation permissions.")
+			log.WithError(err).Warnf("failed to load credentials for cluster %q.", cluster)
+			continue
 		}
+		res[cluster] = clusterCreds
```

Comment:
with this code, if anything in `extractKubeCreds` fails, the k8s cluster will not be registered by `kubernetes_service`.
for an admin to resolve the problem, they need to fix the underlying issue (permissions, firewall, etc) and restart the `kubernetes_service`.

it would be nicer to register the k8s cluster (even if it isn't fully operational) and have it recover without any restarts once the underlying issue is resolved.

this means:
- if anything about credential loading fails, return an error and maybe abort statup - these are local issues in configuration
- if impersonation permission check fails, log a warning, but add `clusterCreds` to `res[cluster` and proceed

Reply:
Done